File: anagrams.cpp
#include "leetcode.h"
class Solution {
public:
    vector<string> anagrams(vector<string> &strs) {
        vector<string> result;
        map<string, vector<string> > m;
        string temp;
        for (int i=0; i < strs.size(); ++i) {
            temp = strs[i];
            sort(temp.begin(), temp.end());
            if (m.find(temp) == m.end()) {
                //m[temp] = new vector<string>(1, strs[i]);
                vector<string> a;
                a.push_back(strs[i]);
                m[temp] = a;
            } else {
                m[temp].push_back(strs[i]);
            }
        }
        for (map<string, vector<string> >::iterator it=m.begin(); it != m.end(); it++) {
            if (it->second.size() > 1) {
                result.insert(result.begin(), it->second.begin(), it->second.end());
            }
        }
        return result;
    }
};
int main() {

    return 0;
}

File: anagrams.java
import java.util.*;
public class Solution {
    public ArrayList<String> anagrams(String[] strs) {
        ArrayList<String> result = new ArrayList<String>();
        HashMap<String, ArrayList<String>> m = new HashMap<String, ArrayList<String>>();
        for (int i=0; i < strs.length; ++i) {
            char[] chars = strs[i].toCharArray();
            Arrays.sort(chars);
            String temp = new String(chars);
            if (m.get(temp) == null) {
                ArrayList<String> a = new ArrayList<String>();
                a.add(strs[i]);
                m.put(temp, a);
            } else {
                m.get(temp).add(strs[i]);
            }
        }
        for (ArrayList<String> it: m.values()) {
            if (it.size() > 1) {
                result.addAll(it);
            }
        }
        return result;
    }
}
